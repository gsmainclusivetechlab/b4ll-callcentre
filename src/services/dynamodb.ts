import AWS from 'aws-sdk';

export interface RecordType {
    count?: number;
    recordingUrl?: string;
}

const getClient = () =>
    new AWS.DynamoDB.DocumentClient({
        endpoint: process.env.DYNAMO_HOST,
        region: process.env.AWS_REGION || 'eu-west-2',
    });

export async function getItem(
    id: string
): Promise<AWS.DynamoDB.DocumentClient.AttributeMap> {
    return getClient()
        .get({
            TableName: process.env.TABLE_NAME || '',
            Key: {
                id,
            },
        })
        .promise()
        .then(({ Item }): RecordType => Item || {});
}

export async function putItem(
    id: string,
    data: RecordType
): Promise<AWS.DynamoDB.DocumentClient.PutItemOutput> {
    return getClient()
        .put({
            TableName: process.env.TABLE_NAME || '',
            Item: {
                id,
                ...data,
            },
        })
        .promise();
}
