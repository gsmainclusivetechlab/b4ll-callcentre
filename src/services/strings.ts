import {
    SayLanguage,
    GatherLanguage,
    SayVoice,
} from 'twilio/lib/twiml/VoiceResponse';
import { translations, MessageId } from '../strings';
import formatMessage from 'format-message';

formatMessage.setup({
    translations,
    missingTranslation: 'error', // don't console.warn or throw an error when a translation is missing
});

const supportedLanguages = [
    'en-GB' as const,
    'fr-FR' as const,
    'en-DEV' as const,
];
export type SupportedLanguage = typeof supportedLanguages[number];

export function isSupportedLanguage(
    language: unknown
): language is SupportedLanguage {
    return supportedLanguages.some((l) => l === language);
}

export function getVoiceParams(
    language: SupportedLanguage
): { language: SayLanguage & GatherLanguage; voice: SayVoice } {
    // voice list: https://docs.aws.amazon.com/polly/latest/dg/voicelist.html
    switch (language) {
        case 'en-GB':
        case 'en-DEV':
        default:
            return {
                language: 'en-GB',
                voice: 'Polly.Emma-Neural',
            };
        case 'fr-FR':
            return {
                language,
                voice: 'Polly.Celine',
            };
        // case 'pt-BR':
        //     return {
        //         language,
        //         voice: 'Polly.Vitoria',
        //     };
        // case 'es-ES':
        //     return {
        //         language,
        //         voice: 'Polly.Lucia',
        //     };
    }
}

type Args = Record<string, unknown>;
export function __(id: MessageId, locale: string): string;
export function __(id: MessageId, args: Args, locale: string): string;
export function __(id: string, a: string | Args, b?: string): string {
    const args = typeof a === 'object' ? a : {};
    const locale = isSupportedLanguage(a)
        ? a
        : isSupportedLanguage(b)
        ? b
        : 'en-GB';

    return formatMessage({ id, default: id }, args, locale);
}
