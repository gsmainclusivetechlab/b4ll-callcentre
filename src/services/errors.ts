import { APIGatewayProxyHandler, APIGatewayProxyEvent } from 'aws-lambda';
import {
    getVoiceParams,
    __,
    isSupportedLanguage,
    SupportedLanguage,
} from './strings';
import { twiml } from 'twilio';
import VoiceResponse from 'twilio/lib/twiml/VoiceResponse';
import { getItem, RecordType } from './dynamodb';
import qs from 'querystring';

export interface ParsedRequest {
    language: SupportedLanguage;
    user: RecordType;
    event: APIGatewayProxyEvent;
}

async function parseRequest(
    event: APIGatewayProxyEvent
): Promise<ParsedRequest> {
    const language = event.pathParameters?.lang;
    if (!isSupportedLanguage(language)) {
        throw new Error('Unsupported language');
    }
    const body = qs.parse(event.body || '');
    const Caller = body.Caller || event.queryStringParameters?.Caller;
    if (typeof Caller !== 'string' || Caller.length < 3) {
        throw new Error('Unable to identify caller');
    }
    const user = await getItem(Caller);
    return {
        language,
        user,
        event,
    };
}

export const safeHandle = (
    f: (e: ParsedRequest) => Promise<VoiceResponse>
): APIGatewayProxyHandler => {
    const safeHandler = async function safelyHandledFunction(
        e: APIGatewayProxyEvent
    ) {
        try {
            const result = await parseRequest(e);
            const response = await f(result);
            return {
                statusCode: 200,
                headers: {
                    'Content-Type': 'text/xml',
                },
                body: response.toString(),
            };
        } catch (err) {
            const maybeLang = e.pathParameters?.lang;
            const language = isSupportedLanguage(maybeLang)
                ? maybeLang
                : 'en-GB';
            const response = new twiml.VoiceResponse();
            const message =
                typeof err === 'string'
                    ? err
                    : typeof err.message === 'string'
                    ? err.message
                    : '';
            response.say(
                getVoiceParams(language),
                __('error', { message }, language)
            );
            return {
                statusCode: 500,
                headers: {
                    'Content-Type': 'text/xml',
                },
                body: response.toString(),
            };
        }
    };
    safeHandler.orig = f;
    return safeHandler;
};
