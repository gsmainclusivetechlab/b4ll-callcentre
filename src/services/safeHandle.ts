import {
    APIGatewayProxyHandler,
    APIGatewayProxyEvent,
    APIGatewayProxyResult,
} from 'aws-lambda';
import {
    getVoiceParams,
    __,
    isSupportedLanguage,
    SupportedLanguage,
} from './strings';
import { twiml } from 'twilio';
import { getAccountItem, RecordType } from './dynamodb';
import qs from 'querystring';
import * as jwt from 'jsonwebtoken';
import {
    HandlerParams,
    handleVerification,
    AuthCookie,
    VerificationState,
} from './auth';

export interface ParsedRequest {
    language: SupportedLanguage;
    user: RecordType;
    event: APIGatewayProxyEvent;
    auth: AuthCookie;
}

async function parseRequest(
    event: APIGatewayProxyEvent
): Promise<ParsedRequest> {
    console.log('parsing', event);
    const language = event.pathParameters?.lang;
    if (!isSupportedLanguage(language)) {
        throw new Error('Unsupported language');
    }
    const body = qs.parse(event.body || '');
    const callerKey =
        body.Direction === 'outbound-api' ||
        event.queryStringParameters?.Direction === 'outbount-api'
            ? 'Called'
            : 'Caller';
    const Caller = body[callerKey] || event.queryStringParameters?.[callerKey];
    if (typeof Caller !== 'string' || Caller.length < 3) {
        throw new Error('Unable to identify caller');
    }
    const user = await getAccountItem(Caller);

    // By default, the user is unauthenticated
    let auth: AuthCookie = {
        sub: Caller,
        state: user.isEnrolled
            ? VerificationState.REGISTERED
            : VerificationState.NOT_ENROLLED,
    };
    const cookie = event.headers.Cookie?.split(/;\s?/)
        ?.map((x) => x.split('='))
        ?.find(([name]) => name === 'Auth')?.[1];
    if (cookie) {
        try {
            if (!process.env.JWT_KEY)
                throw new Error('No JWT Key found in environment');
            auth = jwt.verify(cookie, process.env.JWT_KEY, {
                subject: Caller,
            }) as AuthCookie;
        } catch (e) {
            console.log(e);
            // ignore the failure and treat this as an unauthorized user
        }
    }
    console.log('parse result ', { language, user, event, auth });

    // normalise event path
    if (event.requestContext) {
        event.path = event.requestContext.path;
        (event as any).domainName = event.requestContext.domainName;
    }

    return {
        language,
        user,
        event,
        auth,
    };
}

export const safeHandle = (
    f: (e: ParsedRequest) => Promise<{ toString: () => string }>,
    params: HandlerParams = {}
): APIGatewayProxyHandler => {
    const safeHandler = async function safelyHandledFunction(
        e: APIGatewayProxyEvent
    ) {
        try {
            const request = await parseRequest(e);
            if (params.requireVerification) {
                const result = await handleVerification(request, params);
                if (result) return result;
            }

            const response = await f(request);
            return serialise(response, 200);
        } catch (err) {
            const maybeLang = e.pathParameters?.lang;
            const language = isSupportedLanguage(maybeLang)
                ? maybeLang
                : 'en-GB';
            if (typeof err === 'object') {
                return serialise(err, err.statusCode || 500);
            }
            const response = new twiml.VoiceResponse();
            const message =
                typeof err === 'string'
                    ? err
                    : typeof err.message === 'string'
                    ? err.message
                    : '';
            response.say(
                getVoiceParams(language),
                __('error', { message }, language)
            );
            return serialise(response, 500);
        }
    };
    safeHandler.orig = f;
    return safeHandler;
};

function serialise(r: unknown, statusCode: number): APIGatewayProxyResult {
    console.log('Returning ', r);
    if (typeof r === 'object' && r !== null) {
        if (r instanceof twiml.VoiceResponse) {
            return {
                statusCode,
                headers: {
                    'Content-Type': 'text/xml',
                },
                body: r.toString(),
            };
        }
    }
    return {
        statusCode,
        headers: {
            'Content-Type': 'text/json',
            'access-control-allow-origin': '*',
        },
        body: JSON.stringify(r),
    };
}
