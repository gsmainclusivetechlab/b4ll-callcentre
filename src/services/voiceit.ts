import { VoiceIt } from 'voiceit2-nodejs';
import axios from 'axios';
import { SupportedLanguage } from './strings';

const auth = {
    username: process.env.VOICEIT_API_KEY || '',
    password: process.env.VOICEIT_API_TOKEN || '',
};
const api = axios.create({
    baseURL: 'https://api.voiceit.io/',
    auth,
});

export const AVAILABLE_VOICEPRINTS = {
    'en-GB': {
        Tomorrow: 'Never forget tomorrow is a new day',
        Zoo: 'Zoos are filled with small and large animals',
        Telecom: 'please log in to my telecom services account',
    },
    'en-DEV': {
        // These will get mocked anyway so we can use whatever
        Tomorrow: 'Never forget tomorrow is a new day',
        Zoo: 'Zoos are filled with small and large animals',
        Telecom: 'please log in to my telecom services account',
    },
    'fr-FR': {
        Tomorrow: "N'oubliez jamais que demain est un nouveau jour",
        //  TODO: add voiceprints after they are set up in VoiceIt
    },
};

export type LegalVoicePrint<
    L extends SupportedLanguage
> = keyof typeof AVAILABLE_VOICEPRINTS[L];

export function isLegalPhraseKey<L extends SupportedLanguage>(
    language: L,
    phraseKey: unknown
): phraseKey is LegalVoicePrint<L> {
    if (typeof phraseKey !== 'string') return false;
    return phraseKey in AVAILABLE_VOICEPRINTS[language];
}

export async function createUser(): Promise<VoiceIt.Response<VoiceIt.User>> {
    const result = await api.post('/users');
    return result.data;
}

export async function enrolUser<L extends SupportedLanguage>(
    userId: string,
    language: L,
    enrolment: {
        phrase: keyof typeof AVAILABLE_VOICEPRINTS[L];
        recordingUrl: string;
    }
): Promise<VoiceIt.Response<VoiceIt.VoiceEnrolment>> {
    const matches = AVAILABLE_VOICEPRINTS[language] as Record<string, string>;
    const phrase = matches[enrolment.phrase as string];
    const result = await api.post('/enrollments/voice/byUrl', {
        userId,
        phrase,
        fileUrl: enrolment.recordingUrl,
        contentLanguage: 'en-US',
        // TODO: Restore language when we've got a full account
        // contentLanguage: language,
    });
    return result.data;
}

export async function verifyUser<L extends SupportedLanguage>(
    userId: string,
    language: L,
    enrolment: {
        phrase: keyof typeof AVAILABLE_VOICEPRINTS[L];
        recordingUrl: string;
    }
): Promise<VoiceIt.Response<{ confidence: number }>> {
    const matches = AVAILABLE_VOICEPRINTS[language] as Record<string, string>;
    const phrase = matches[enrolment.phrase as string];
    const result = await api.post('/verification/voice/byUrl', {
        userId,
        phrase,
        fileUrl: enrolment.recordingUrl,
        contentLanguage: 'en-US',
        // TODO: Restore language when we've got a full account
        // contentLanguage: language,
    });
    return result.data;
}
