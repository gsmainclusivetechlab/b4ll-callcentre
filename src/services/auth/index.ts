import { APIGatewayProxyResult } from 'aws-lambda';
import { getVoiceParams, __ } from '../strings';
import { twiml } from 'twilio';
import { requestVerification } from './requestVerification';
import { checkVerification } from './checkVerification';
import { handleEnrolment } from './handleEnrolment';
import { requestEnrolment } from './requestEnrolment';
import { ParsedRequest } from '../safeHandle';
import * as jwt from 'jsonwebtoken';
import {
    VoiceItEnrolmentData,
    VoiceItVerificationData,
} from '../../engine/voiceit/provider';

export enum VerificationState {
    NOT_ENROLLED = 0,
    ENROLMENT_REQUESTED = 1,
    REGISTERED = 2,
    AUTHENTICATION_REQUESTED = 3,
    AUTHENTICATED = 4,
    DEACTIVATED = 5,
}
interface BaseCookie {
    /** caller phone number */
    sub: string;
    exp?: number;
}
interface PlainCookie extends BaseCookie {
    state:
        | VerificationState.AUTHENTICATED
        | VerificationState.NOT_ENROLLED
        | VerificationState.REGISTERED
        | VerificationState.DEACTIVATED;
}
interface EnrolmentRequestedCookie extends BaseCookie {
    state: VerificationState.ENROLMENT_REQUESTED;
    /** phrase that the user was asked to record */
    phrase: string;
    req: VoiceItEnrolmentData;
}
interface AuthenticationRequestedCookie extends BaseCookie {
    state: VerificationState.AUTHENTICATION_REQUESTED;
    req: VoiceItVerificationData;
}

export type AuthCookie =
    | PlainCookie
    | AuthenticationRequestedCookie
    | EnrolmentRequestedCookie;

export function makeCookieHeader(cookie: AuthCookie): Record<string, string> {
    const signed = jwt.sign(cookie, process.env.JWT_KEY || '', {
        expiresIn: '30s',
    });
    return {
        'Set-Cookie': `Auth=${signed}`,
    };
}

export interface HandlerParams {
    requireVerification?: boolean;
    allowEnrolment?: boolean;
    allowDeactivated?: boolean;
    loginRedirect?: { method: string; target: string };
}

export async function handleVerification(
    request: ParsedRequest,
    params: HandlerParams
): Promise<null | APIGatewayProxyResult> {
    const { auth, language } = request;
    console.log('Verifying, user is in state ', auth);
    switch (auth?.state) {
        default:
        case VerificationState.NOT_ENROLLED:
            if (params.allowEnrolment === false) break;
            return requestEnrolment(request);
        case VerificationState.ENROLMENT_REQUESTED:
            return handleEnrolment(request, params.loginRedirect);
        case VerificationState.REGISTERED:
            return requestVerification(request);
        case VerificationState.AUTHENTICATION_REQUESTED:
            return checkVerification(request, params.loginRedirect);
        case VerificationState.AUTHENTICATED:
            return null;
        case VerificationState.DEACTIVATED:
            if (params.allowDeactivated === true) return null; // proceed to the original function handler
            break; // proceed to the unauthorised handler below
    }

    // currently the only way to arrive here is if enrolment is required but not allowed on the handler
    const response = new twiml.VoiceResponse();
    response.say(getVoiceParams(language), __('unauthorised', language));
    return {
        statusCode: 200,
        headers: {
            'Content-Type': 'text/xml',
        },
        body: response.toString(),
    };
}
