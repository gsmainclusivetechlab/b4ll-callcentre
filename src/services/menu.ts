import { twiml } from 'twilio';
import { getVoiceParams, __ } from './strings';
import { ParsedRequest } from './safeHandle';
import querystring from 'querystring';
import { MessageId } from '../strings';
import VoiceResponse from 'twilio/lib/twiml/VoiceResponse';

export interface MenuOption {
    triggers: string[];
    description: MessageId | ((r: ParsedRequest) => Promise<MessageId>);
    handler: (r: ParsedRequest) => Promise<VoiceResponse>;
}

export async function menuToHandler(
    menu: MenuOption[],
    { event, language, user, auth }: ParsedRequest,
    redirect?: string
): Promise<VoiceResponse> {
    const { Digits, SpeechResult } = querystring.parse(event.body || '');
    const answer = Digits || SpeechResult || null;
    if (typeof answer === 'string') {
        const selected = menu.find(
            (option, i) =>
                answer === '' + (i + 1) || option.triggers.indexOf(answer) >= 0
        );
        if (selected) {
            return selected.handler({ language, user, event, auth });
        }
    }

    // redirect to root and we can try again
    const response = new twiml.VoiceResponse();
    response.say(getVoiceParams(language), __('did-not-understand', language));
    if (redirect) response.redirect({ method: 'GET' }, redirect);
    return response;
}

export async function menuToGather(
    response: VoiceResponse,
    request: ParsedRequest,
    menu: MenuOption[]
): Promise<VoiceResponse.Gather> {
    const voice = getVoiceParams(request.language);
    const gather = response.gather({
        input: ['dtmf'],
        method: 'POST',
        language: voice.language,
        hints: menu.flatMap((o) => o.triggers).join(','),
        numDigits: 1,
        timeout: 5,
    });

    // TODO: consider better behaviour in empty menu case and > 9 items case
    for (let i = 0; i < Math.min(menu.length, 9); i++) {
        const { description: maybeDesc } = menu[i];
        const description: MessageId =
            typeof maybeDesc === 'string'
                ? maybeDesc
                : await maybeDesc(request);
        gather.say(voice, __(description, { index: i + 1 }, request.language));
    }

    return gather;
}
