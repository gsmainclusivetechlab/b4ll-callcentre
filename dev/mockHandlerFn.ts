/* istanbul ignore file */

import { safeHandle, ParsedRequest } from '../src/services/safeHandle';

type DeepPartial<T> = {
    [P in keyof T]?: DeepPartial<T[P]>;
};
type Handler = Parameters<typeof safeHandle>[0];
type TestInput = Omit<ParsedRequest, 'event' | 'auth'> & {
    event?: DeepPartial<ParsedRequest['event']>;
    auth?: ParsedRequest['auth'];
};
type HandlerResult = ReturnType<Handler>;

/**
 * Undoes the effect of the `safeHandle` function, and mocks the event parameter
 * @param handler
 */
export function mockHandlerFn(
    handler: unknown
): (input: TestInput) => HandlerResult {
    if (
        typeof handler === 'function' &&
        handler.name === 'safelyHandledFunction' &&
        'orig' in handler
    ) {
        const orig = (handler as Record<string, unknown>)['orig'];
        if (typeof orig === 'function') {
            return ({ event, ...params }) =>
                orig({ ...params, event: event || {} });
        }
    }
    throw new Error("Can't mock handlers which are not safely handled");
}
