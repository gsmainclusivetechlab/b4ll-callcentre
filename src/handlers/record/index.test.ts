import { handler as orig } from '.';
import qs from 'querystring';
import { mockHandlerFn } from '../../../dev/mockHandlerFn';
import { getItem } from '../../services/dynamodb';

const handler = mockHandlerFn(orig);
describe('Recording', () => {
    it('should return well-formed XML to initial request', async () => {
        const result = await handler({
            pathParameters: { lang: 'en-GB' },
            queryStringParameters: { Caller: '+7777777' },
            httpMethod: 'GET',
        });
        expect(result).toMatchObject({
            statusCode: 200,
            headers: {
                'Content-Type': 'text/xml',
            },
            body: expect.any(String),
        });
        expect(result.body).toMatchSnapshot();
    });

    it('should return well-formed XML to subsequent response', async () => {
        const initialItem = await getItem('+7777777');
        expect(initialItem.recordingUrl).not.toEqual(
            'https://my-file-server/voice.wav'
        );
        const result = await handler({
            pathParameters: { lang: 'en-GB' },
            body: qs.stringify({
                Caller: '+7777777',
                RecordingUrl: 'https://my-file-server/voice.wav',
            }),
            httpMethod: 'POST',
        });
        expect(result).toMatchObject({
            statusCode: 200,
            headers: {
                'Content-Type': 'text/xml',
            },
            body: expect.any(String),
        });
        expect(result.body).toMatchSnapshot();
        const savedItem = await getItem('+7777777');
        expect(savedItem.recordingUrl).toEqual(
            'https://my-file-server/voice.wav'
        );
    });

    test.each`
        error                                  | data
        ${'Unable to identify caller'}         | ${{}}
        ${'Unable to identify caller'}         | ${{ queryStringParameters: { Caller: '+1' } }}
        ${'Unsupported language'}              | ${{ queryStringParameters: { Caller: '+7777777' } }}
        ${'Could not retrieve recording URL.'} | ${{ pathParameters: { lang: 'en-GB' }, body: qs.stringify({ Caller: '+7777777' }), httpMethod: 'POST' }}
    `('should gracefully reject: $error', async ({ data }) => {
        const result = await handler(data);
        expect(result.body).toMatchSnapshot();
    });
});
