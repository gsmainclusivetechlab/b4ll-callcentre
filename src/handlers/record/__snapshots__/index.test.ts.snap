// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recording should gracefully reject: Could not retrieve recording URL. 1`] = `
<?xml version="1.0" encoding="UTF-8"?>
<Response>
    <Say language="en-GB" voice="Polly.Amy">
        An error occurred. Could not retrieve recording URL.
    </Say>
</Response>
`;

exports[`Recording should gracefully reject: Unable to identify caller 1`] = `
<?xml version="1.0" encoding="UTF-8"?>
<Response>
    <Say language="en-GB" voice="Polly.Amy">
        An error occurred. Unable to identify caller
    </Say>
</Response>
`;

exports[`Recording should gracefully reject: Unable to identify caller 2`] = `
<?xml version="1.0" encoding="UTF-8"?>
<Response>
    <Say language="en-GB" voice="Polly.Amy">
        An error occurred. Unable to identify caller
    </Say>
</Response>
`;

exports[`Recording should gracefully reject: Unsupported language 1`] = `
<?xml version="1.0" encoding="UTF-8"?>
<Response>
    <Say language="en-GB" voice="Polly.Amy">
        An error occurred. Unsupported language
    </Say>
</Response>
`;

exports[`Recording should return well-formed XML to initial request 1`] = `
<?xml version="1.0" encoding="UTF-8"?>
<Response>
    <Say language="en-GB" voice="Polly.Amy">
        Please record your message, then press the hash key.
    </Say>
    <Record method="POST" finishOnKey="#" playBeep="true" trim="trim-silence" maxLength="5"/>
</Response>
`;

exports[`Recording should return well-formed XML to subsequent response 1`] = `
<?xml version="1.0" encoding="UTF-8"?>
<Response>
    <Say language="en-GB" voice="Polly.Amy">
        Awesome! Here's what you sounded like:
    </Say>
    <Play>
        https://my-file-server/voice.wav
    </Play>
</Response>
`;
